From 58f31eee68eb8b9b181de4fbfa38571721fdd01e Mon Sep 17 00:00:00 2001
From: Tim Stahlhut <stahta01@gmail.com>
Date: Thu, 9 Feb 2023 00:25:21 -0500
Subject: msys2 fixes for tools/common.mk

1. Create ISMSYS_MINGW64 flag to indicate building under MSys2 with an
   Compiler with an mingw64 based triplet.
2. Add MSys2 settings to CMAKE_COMMON.
3. Add CREATE_TAR_FLAGS to help create tars under MSys2.
4. Use "MSYS Makefiles"
---
 Make.inc             | 12 ++++++++++++
 deps/tools/common.mk | 17 ++++++++++++++++-
 2 files changed, 28 insertions(+), 1 deletion(-)

diff --git a/Make.inc b/Make.inc
index ce2590b2f5..7ff4972847 100644
--- a/Make.inc
+++ b/Make.inc
@@ -374,6 +374,18 @@ export PKG_CONFIG_LIBDIR = $(JULIAHOME)/usr/lib/pkgconfig
 # Figure out OS and architecture
 BUILD_OS := $(shell uname)
 
+# Set ISMSYS_MINGW64 to 0 or 1
+OSNAME := $(shell uname -o | tr '[:lower:]' '[:upper:]')
+ISMSYS_MINGW64 := 0
+ifneq (,$(findstring MSYS,$(OSNAME)))
+TRIPLET := $(shell $(CC) -dumpmachine)
+ifneq (,$(findstring w64,$(TRIPLET)))
+ifneq (,$(findstring mingw,$(TRIPLET))) 
+override ISMSYS_MINGW64 := 1
+endif
+endif
+endif
+
 ifneq (,$(findstring CYGWIN,$(BUILD_OS)))
 XC_HOST ?= $(shell uname -m)-w64-mingw32
 endif
diff --git a/deps/tools/common.mk b/deps/tools/common.mk
index 642528376d..f541c7e195 100644
--- a/deps/tools/common.mk
+++ b/deps/tools/common.mk
@@ -28,8 +28,12 @@ endif
 ifneq ($(VERBOSE), 0)
 CMAKE_COMMON += -DCMAKE_VERBOSE_MAKEFILE=ON
 endif
+ifeq (1,$(ISMSYS_MINGW64))
+CMAKE_COMMON += -DCMAKE_C_COMPILER="$$(cygpath -m $$(which $(CC_BASE)))"
+else
 # The call to which here is to work around https://cmake.org/Bug/view.php?id=14366
 CMAKE_COMMON += -DCMAKE_C_COMPILER="$$(which $(CC_BASE))"
+endif
 ifneq ($(strip $(CMAKE_CC_ARG)),)
 CMAKE_COMMON += -DCMAKE_C_COMPILER_ARG1="$(CMAKE_CC_ARG)"
 endif
@@ -48,7 +52,11 @@ endif
 ifeq ($(CMAKE_GENERATOR),Ninja)
 CMAKE_GENERATOR_COMMAND := -G Ninja
 else ifeq ($(CMAKE_GENERATOR),make)
+ifeq (1,$(ISMSYS_MINGW64))
+CMAKE_GENERATOR_COMMAND := -G "MSYS Makefiles"
+else
 CMAKE_GENERATOR_COMMAND := -G "Unix Makefiles"
+endif
 else
 $(error Unknown CMake generator '$(CMAKE_GENERATOR)'. Options are 'Ninja' and 'make')
 endif
@@ -160,11 +168,18 @@ reinstall-$(strip $1):
 	+$$(MAKE) stage-$(strip $1)
 	+$$(MAKE) install-$(strip $1)
 
+ifeq (1,$(ISMSYS_MINGW64))
+# MSys2 has issues with symbolic links
+CREATE_TAR_FLAGS := --dereference -czf
+else
+CREATE_TAR_FLAGS := -czf
+endif
+
 $$(build_staging)/$2.tgz: $$(BUILDDIR)/$2/build-compiled
 	rm -rf $$(build_staging)/$2
 	mkdir -p $$(build_staging)/$2$$(build_prefix)
 	$(call $3,$$(BUILDDIR)/$2,$$(build_staging)/$2,$4)
-	cd $$(build_staging)/$2$$(build_prefix) && $$(TAR) -czf $$@.tmp .
+	cd $$(build_staging)/$2$$(build_prefix) && $$(TAR) $$(CREATE_TAR_FLAGS) $$@.tmp .
 	rm -rf $$(build_staging)/$2
 	mv $$@.tmp $$@
 
-- 
